subsystem/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+        let subsys_name = self.subsystem_name;
+        // Shutdown-related channels
+        let shutting_down_tx = self.manager.shutting_down_tx.clone();
+        let shutdown_rq = ShutdownRequest(self.manager.shutdown_request_tx.subscribe());
+        // Call related channels
+        let (action_tx, action_rx) = mpsc::channel(self.call_queue_capacity);
+        let call_rq = CallRequest(action_rx);
+
+        task::spawn(async move {
+            log::info!("Subsystem {}/{} started", manager_name, subsys_name);
+
+            // Perform the subsystem task.
+            subsystem(call_rq, shutdown_rq).await;
+
+            // Signal the intent to shut down to the other parts of the application.
+            shutting_down_tx.send(()).await.expect("Subsystem outlived the manager!?");
+
+            log::info!("Subsystem {}/{} terminated", manager_name, subsys_name);
+
+            // Close the channel to signal the completion of the shutdown.
+            std::mem::drop(shutting_down_tx);
+        });
+
+        Subsystem::new(action_tx)
+    }
+
+    /// Start a passive subsystem.
+    ///
+    /// A passive subsystem does not interact with the environment on its own. It only serves calls
+    /// from other subsystems.
+    pub fn start_passive<T: 'static + Send>(self, mut obj: T) -> Subsystem<T> {
+        self.start(|mut call_rq, mut shutdown_rq| async move {
+            loop {
+                tokio::select! {
+                    () = shutdown_rq.recv() => { break; }
+                    call = call_rq.recv() => { call(&mut obj); }
+                }
+            }
+        })
+    }
+}
+
+/// Internal action type sent in the channel.
+type Action<T> = Box<dyn FnOnce(&mut T) + Send>;
+
+/// Call request
+pub struct CallRequest<T>(mpsc::Receiver<Action<T>>);
+
+impl<T: 'static + Send> CallRequest<T> {
+    /// Receive an external call to this subsystem.
+    pub async fn recv(&mut self) -> Action<T> {
+        match self.0.recv().await {
+            // We have a call, return it
+            Some(action) => action,
+            // All handles to this subsystem dropped, suspend call handling.
+            None => std::future::pending().await,
+        }
+    }
+}
+
+/// Shutdown request
+pub struct ShutdownRequest(broadcast::Receiver<()>);
+
+impl ShutdownRequest {
+    /// Receive a shutdown request.
+    pub async fn recv(&mut self) {
+        match self.0.recv().await {
+            Err(broadcast::error::RecvError::Lagged(_)) => {
+                panic!("Multiple shutdown broadcast requests issued")
+            }
+            Err(broadcast::error::RecvError::Closed) => {
+                panic!("Shutdown channel sender closed prematurely")
+            }
+            Ok(()) => (),
+        }
+    }
+}
+
+/// Subsystem handle.
+///
+/// This allows the user to interact with the subsystem from the outside. Currently, it only
+/// supports calling functions on the subsystem.
+pub struct Subsystem<T> {
```

Comment:
While this type represents an interface to a subsystem, it is not in itself a subsystem. I am considering renaming it to `Handle` so it would be referred to as `subsystem::Handle` from other crates. Would you consider it an improvement?

Reply:
There was no feedback so I decided to go ahead with this for the time of being. Any objections let me know.